### PR TITLE
Merge Crafting Jobs

### PR DESCRIPTION
**Pull request by pashabou**
Allow jobs to be merged on CPUs already crafting the target item.

https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/assets/6406169/53fa870c-76d2-4755-8533-5c7b5b79b636

### CODE REVIEW

**Comment by budak7273** [comment]
Can you avoid the merge (intentionally create a separate job) by selecting a different CPU, or will it always merge? The video doesn't showcase this.

**Comment by Quetz4l** [comment]
Will there be a hotkey for auto-mergening? Because large recipes can take a long time to process, and some component might disappear. mb ctrl + shift +next button?

**Comment by pashabou** [comment]
> Can you avoid the merge (intentionally create a separate job) by selecting a different CPU, or will it always merge? The video doesn't showcase this.

Yes, you can still select any free CPU like before. Selection defaults to an occupied CPU if merging is available though.

> Will there be a hotkey for auto-mergening? Because large recipes can take a long time to process, and some component might disappear. mb ctrl + shift +next button?

I know what you mean by components disappearing between typing the amount to craft and and hitting "Start" on the confirmation screen, but I wasn't going to add a hotkey. Is there one for the non-merge case?

**Comment by StaffiX** [comment]
clicking the next button via holding ctrl skips the page where it lists the items it's going to craft and use

**Comment by pashabou** [comment]
> clicking the next button via holding ctrl skips the page where it lists the items it's going to craft and use

Looks like it's the shift key for me, just tested and this works already @Quetz4l .

**Comment by ehami** [comment]
What happens when you are merging a new crafting request with an old one that was initially created with a now-removed pattern?

**Comment by ivelieu** [comment]
Behavior with crafting cards should be toggleable, so that the player can choose whether automatic crafting card / level maintainer requests are combined into an existing CPU or not.

**Comment by pashabou** [comment]
> What happens when you are merging a new crafting request with an old one that was initially created with a now-removed pattern?

Just tested this, in-progress jobs are aware of pattern changes, so for example if I move a pattern for a Molecular Assembler to another interface it will still work and use the new interface. When the crafting tree is calculated, it will take into account the available patterns, so if you remove a pattern to make an intermediate after starting the job the task for that intermediate will get stuck on "Scheduled". If you create a second crafting tree (for the to-be-merged job) it will be as normal, taking into account the currently available patterns, and potentially reporting that a certain intermediate is missing from the craft as it no longer knows how to make that (again, as normal).

TL;DR: It just works™

> Behavior with crafting cards should be toggleable, so that the player can choose whether automatic crafting card / level maintainer requests are combined into an existing CPU or not.

I'll consider this as part of a future change I want to make to give more control over crafting card & maintainer job placement.

**Comment by MCTBL** [comment]
LGTM, Not sure if use in GTNH has any other bugs. At least I didn't encounter any bugs in the dev env.
Need more test.

**Comment by ehami** [comment]
How does it handle cpu storage limitations? It shouldn’t be able to exceed the cap between old+new. Ideally it’d wouldn’t merge in that case (ie, just create a new job on a new cpu) if it doesn’t fit alongside the old task on the old cpu).

The ideal solution of migrating the job from one cpu that could fit old+new as a merged job is probably not worth the implementation effort.

**Comment by MCTBL** [comment]
> How does it handle cpu storage limitations? It shouldn’t be able to exceed the cap between old+new. Ideally it’d wouldn’t merge in that case (ie, just create a new job on a new cpu) if it doesn’t fit alongside the old task on the old cpu).
> 
> The ideal solution of migrating the job from one cpu that could fit old+new as a merged job is probably not worth the implementation effort.

when i test it will add old usage and new usage then check if its lower than the capacity of that cpu. If is lower than the capacity that means it can merge, Otherwise it can't.

**Comment by MCTBL** [comment]
btw, may make cpu show up How many capacity it already in usea is better?

**Comment by pashabou** [comment]
Thanks for the suggestion @MCTBL, I made sure the values in the tooltip reflect what's going on in a merged job (step counts & used / total crafting storage).

**Comment by tth05** [comment]
Crash probably caused by this PR https://discord.com/channels/181078474394566657/181078474394566657/1263472857128632381
[crash-2024-07-18_14.26.23-server.txt](https://github.com/user-attachments/files/16283361/crash-2024-07-18_14.26.23-server.txt)

File: src/main/java/appeng/container/implementations/ContainerCraftAmount.java [hunk 1]
```diff
@@ -76,5 +76,8 @@ public void setItemToCraft(@Nonnull final IAEItemStack itemToCreate) {
 
     public void openConfirmationGUI(EntityPlayer player, TileEntity te) {
         Platform.openGUI(player, te, this.getOpenContext().getSide(), GuiBridge.GUI_CRAFTING_CONFIRM);
+        if (player.openContainer instanceof ContainerCraftConfirm ccc) {
+            ccc.setItemToCraft(this.itemToCreate);
+        }
```
Comment (Ruling-0) [inline src/main/java/appeng/container/implementations/ContainerCraftAmount.java:81]: It appears this is invoked from the set amount window, so another method (such as ctrl-clicking the NEI question mark) which bypasses this window while a craft is occurring likely results in it checking to merge with an empty ByteBuf, leading to the crash described.

Comment (pashabou) [reply src/main/java/appeng/container/implementations/ContainerCraftAmount.java:81]: Huge thanks for narrowing this down for me, this was exactly it. Didn't think that this would be called from other mods but now I know!

**Comment by pashabou** [comment]
Thanks @tth05 and @Ruling-0 for helping me to reproduce the issue. CTRL+Click-ing in NEI while a crafting job for an intermediate is already running won't crash the game anymore.

**Comment by pashabou** [comment]
Three new PRs for everywhere a `ContainerCraftConfirm` GUI is created and navigated to. Tested and fixes crashes with wireless terminals. Hopefully that should cover this issue but if there's more crashes send em my way.

**Comment by Ruling-0** [comment]
Nice work, glad I could help. Newer nightly (563) I can confirm the ctrl-click crash is solved, but myself and another managed to crash using the fluid pattern terminal - it looks like your PR to that mod failed the git check and that fix might not be deployed in a nightly.

**Comment by evgengoldwar** [comment]
Checked on the latest nightly, no crashes.

**Comment by awolosewicz** [comment]
Checked on new latest nightly, wireless crafting terminal crashes

[crash-2024-07-26_15.44.40-server.txt](https://github.com/user-attachments/files/16396557/crash-2024-07-26_15.44.40-server.txt)

**Comment by Ruling-0** [comment]
Interestingly, I did manage to crash on latest nightly. It's Wireless Crafting Terminal, just the usual steps of starting 1 craft then trying to start a second

[crash-2024-07-26_15.05.02-client.txt](https://github.com/user-attachments/files/16397796/crash-2024-07-26_15.05.02-client.txt)

File: src/main/java/appeng/api/networking/crafting/ICraftingCPU.java [hunk 2]
```diff
@@ -36,6 +36,11 @@ public interface ICraftingCPU extends IBaseMonitor<IAEItemStack> {
      */
     long getAvailableStorage();
 
+    /**
+     * @return the storage used by an active crafting job in bytes
+     */
+    long getUsedStorage();
```
Comment (eigenraven) [inline src/main/java/appeng/api/networking/crafting/ICraftingCPU.java:42]: This needs to have a `default` implementation to preserve backwards API compatibility for any old addons

**Review by eigenraven — COMMENTED** [review]
Overall looks pretty good, please make sure that this code can work with old, unupdated AE2 addons without any issues - GTNH AE2 is used outside of GTNH in other packs

File: src/main/java/appeng/api/networking/crafting/ICraftingCPU.java (hunk 2 above)
Comment (pashabou) [reply src/main/java/appeng/api/networking/crafting/ICraftingCPU.java]: Done, thanks for pointing this out

**Review by eigenraven — APPROVED** [review]
